geode-cq/src/distributedTest/java/org/apache/geode/internal/cache/tier/sockets/DurableClientCQClusterRestartDUnitTest.java

```diff
@@ -0,0 +1,315 @@
+  private VM locator;
+  private VM client;
+  private File locatorLog;
+  private File serverLog;
+  private File restartLocatorLog;
+  private File restartServerLog;
+  private int locatorPort;
+  private String durableClientId;
+  private final String cqName = "cqQuery";
+  private final int numOfInvocations = 5;
+  private final int uniquePort = new UniquePortSupplier().getAvailablePort();
+
+  @Rule
+  public DistributedRule distributedRule = new DistributedRule();
+
+  @Rule
+  public CacheRule cacheRule = new CacheRule();
+
+  @Rule
+  public ClientCacheRule clientCacheRule = new ClientCacheRule();
+
+  @Rule
+  public SerializableTemporaryFolder temporaryFolder = new SerializableTemporaryFolder();
+
+  @Rule
+  public SerializableTestName testName = new SerializableTestName();
+
+  @Rule
+  public DistributedExecutorServiceRule executorServiceRule = new DistributedExecutorServiceRule();
+
+  @Before
+  public void setup() throws Exception {
+    locator = VM.getVM(0);
+    server = VM.getVM(1);
+    client = VM.getVM(2);
+
+    hostName = getHostName();
+    uniqueName = getClass().getSimpleName() + "_" + testName.getMethodName();
+    regionName = uniqueName + "_region";
+    locatorLog = new File(temporaryFolder.newFolder(uniqueName), "locator.log");
+    serverLog = temporaryFolder.getRoot().toPath().resolve("server.log").toFile();
+    restartLocatorLog = temporaryFolder.getRoot().toPath().resolve("locator_restart.log").toFile();
+    restartServerLog = temporaryFolder.getRoot().toPath().resolve("server_restart.log").toFile();
+    durableClientId = uniqueName + "client";
+  }
+
+  @Test
+  public void cqEventsNotLostIfClusterRestarted() {
+    startClusterAndDoFunctionCalls(locatorLog, serverLog, 1);
+
+    // restart cluster and perform another round of function calls.
+    startClusterAndDoFunctionCalls(restartLocatorLog, restartServerLog, 2);
+
+  }
+
+  private void startClusterAndDoFunctionCalls(File locatorLog, File serverLog, int iteration) {
+    locatorPort = locator.invoke(() -> startLocator(locatorLog));
+    server.invoke(() -> createCacheServerAndDiskRegion(serverLog));
+    client.invoke(this::setClientRegion);
+    client.invoke(() -> clientCacheRule.getClientCache().readyForEvents());
+    client.invoke(this::callFunctions);
+    client.invoke(() -> verifyCQListenerInvocations(numOfInvocations * iteration));
+    server.invoke(this::closeCache);
+    server.bounceForcibly();
+    locator.invoke(this::stopLocator);
+  }
+
+  private int startLocator(File locatorLog) throws IOException {
+    InetAddress bindAddress = InetAddress.getByName(hostName);
+    Locator locator =
+        Locator.startLocatorAndDS(locatorPort, locatorLog, bindAddress, new Properties());
+    return locator.getPort();
+  }
+
+  private void createCacheServerAndDiskRegion(File logFile) throws Exception {
+    Properties systemProperties = new Properties();
+    systemProperties.setProperty("gemfire.jg-bind-port", Integer.toString(uniquePort));
```

Comment:
I've never seen this property used before. What's the purpose for including it in this test?

Reply:
The jgroup port is used as part of EventID.membership field to distinguish event id. If a new jgroup port is chosen between the restart, the test will not fail (before this fix). This is to make sure to eliminate the that condition, so if something breaks, it can be captured earlier.